{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# part 1 get bed files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#part1.1:  get position bed from transcript bed +\n",
    "# v0.2 add stop codon\n",
    "def getPosBed_Plus(transBed):\n",
    "    arrTest=transBed\n",
    "    \n",
    "    ## init paras\n",
    "    arrExon=[]\n",
    "    arrUTR=[]\n",
    "    start_codon=[] #one\n",
    "    stop_codon=[] #one\n",
    "    #\n",
    "    for tmp in arrTest:\n",
    "        #arr3=arrTest[i];\n",
    "        if tmp[-1]=='exon':\n",
    "            arrExon.append(tmp)\n",
    "        elif tmp[-1]=='UTR':\n",
    "            arrUTR.append(tmp)\n",
    "        elif tmp[-1]=='start_codon':\n",
    "            start_codon=tmp;\n",
    "        elif tmp[-1]=='stop_codon':\n",
    "            stop_codon=tmp;\n",
    "        #print(arrTest[i])\n",
    "    #print(len(arrExon), len(arrUTR))\n",
    "    \n",
    "    ## maybe have 3UTR\n",
    "    UTR3=[]\n",
    "    UTR5=[]\n",
    "    if len(arrUTR)>0:\n",
    "        for tmp in arrUTR:\n",
    "            if len(start_codon)>0 and tmp[2]<start_codon[1]:\n",
    "                tmp[-1]=\"UTR5\"\n",
    "                UTR5.append(tmp)\n",
    "            elif len(stop_codon)>0 and tmp[1]>=stop_codon[1]:\n",
    "                tmp[-1]=\"UTR3\"\n",
    "                UTR3.append(tmp)\n",
    "    #print(len(UTR5), len(UTR3))\n",
    "    \n",
    "    ## maybe have e3UTR\n",
    "    eUTR3=[]\n",
    "    if len(UTR3)>0:\n",
    "        eUTR3=UTR3[-1].copy()\n",
    "        #\n",
    "        eUTR3[1]= str( int(eUTR3[2])+1 )\n",
    "        eUTR3[2]= str( int(eUTR3[2])+1000 )\n",
    "        eUTR3[6]='extended3UTR'\n",
    "        #print(UTR3[-1])\n",
    "        #print(eUTR3)\n",
    "    ##get arrIntron，maybe\n",
    "    arrIntron=[]\n",
    "    #print(arrTest[0], '\\n')\n",
    "    intron=[]\n",
    "    if len(arrExon)>1: #if only one exon, then no intron.\n",
    "        for i in range(len(arrExon)):\n",
    "            tmp=arrExon[i]\n",
    "            if len(intron)==0: #first exon\n",
    "                intron=tmp.copy()\n",
    "                intron[1]= str( int(intron[2])+1 )\n",
    "                continue;\n",
    "            else:\n",
    "                #2nd\n",
    "                intron[2]=str( int(tmp[1])-1 )\n",
    "                intron[-1]=\"intron\"\n",
    "                arrIntron.append(intron);\n",
    "                #\n",
    "                intron=tmp.copy();\n",
    "                intron[1]= str( int(intron[2])+1 )\n",
    "    #\n",
    "    #print('arrExon=',arrExon)\n",
    "    #print('arrIntron=',arrIntron)\n",
    "\n",
    "    ##get TSS，must have\n",
    "    TSS=arrTest[0].copy()\n",
    "    pos=int(arrTest[0][1])\n",
    "    TSS[2]=str( pos+10 )\n",
    "    TSS[1]=str( pos-10 )\n",
    "    TSS[-1]=\"TSS\"\n",
    "    #print(arrTest[0])\n",
    "    #TSS\n",
    "    \n",
    "    ##get Promoter，must have\n",
    "    Promoter=arrTest[0].copy()\n",
    "    Promoter[1]=str( pos-250 )\n",
    "    Promoter[2]=str( pos-11 )\n",
    "    Promoter[-1]='Promoter'\n",
    "    #print(arrTest[0])\n",
    "    #Promoter\n",
    "    \n",
    "    #get PA, must have\n",
    "    PA=arrTest[0].copy()\n",
    "    pos2=int(arrTest[0][2])\n",
    "    PA[1]=str( pos2-10 )\n",
    "    PA[2]=str( pos2+10 )\n",
    "    PA[-1]='PA'\n",
    "    \n",
    "    # combine\n",
    "    saveSet=[]\n",
    "    saveSet.extend(UTR3) #many\n",
    "    saveSet.append(eUTR3) #1\n",
    "\n",
    "    saveSet.extend(arrExon) #many\n",
    "    saveSet.extend(arrIntron) #many\n",
    "    saveSet.extend(UTR5) #many\n",
    "\n",
    "    saveSet.append(TSS) #1\n",
    "    saveSet.append(Promoter) #1\n",
    "    \n",
    "    saveSet.append(arrTest[0]) #1\n",
    "    \n",
    "    saveSet.append(stop_codon)\n",
    "    saveSet.append(start_codon)\n",
    "    saveSet.append(PA)\n",
    "\n",
    "    return saveSet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['chr1', '70006', '71585', 'OR4F5-202', '0', '+', 'UTR3'],\n",
       " ['chr1', '71586', '72585', 'OR4F5-202', '0', '+', 'extended3UTR'],\n",
       " ['chr1', '65419', '65433', 'OR4F5-202', '0', '+', 'exon'],\n",
       " ['chr1', '65520', '65573', 'OR4F5-202', '0', '+', 'exon'],\n",
       " ['chr1', '69037', '71585', 'OR4F5-202', '0', '+', 'exon'],\n",
       " ['chr1', '65434', '65519', 'OR4F5-202', '0', '+', 'intron'],\n",
       " ['chr1', '65574', '69036', 'OR4F5-202', '0', '+', 'intron'],\n",
       " ['chr1', '65419', '65433', 'OR4F5-202', '0', '+', 'UTR5'],\n",
       " ['chr1', '65520', '65564', 'OR4F5-202', '0', '+', 'UTR5'],\n",
       " ['chr1', '65409', '65429', 'OR4F5-202', '0', '+', 'TSS'],\n",
       " ['chr1', '65169', '65408', 'OR4F5-202', '0', '+', 'Promoter'],\n",
       " ['chr1', '65419', '71585', 'OR4F5-202', '0', '+', 'transcript'],\n",
       " ['chr1', '70006', '70008', 'OR4F5-202', '0', '+', 'stop_codon'],\n",
       " ['chr1', '65565', '65567', 'OR4F5-202', '0', '+', 'start_codon'],\n",
       " ['chr1', '71575', '71595', 'OR4F5-202', '0', '+', 'PA']]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#test\n",
    "arrTestP=[\n",
    "    ['chr1', '65419', '71585', 'OR4F5-202', '0', '+', 'transcript'],\n",
    "    ['chr1', '65419', '65433', 'OR4F5-202', '0', '+', 'exon'],\n",
    "    ['chr1', '65520', '65573', 'OR4F5-202', '0', '+', 'exon'],\n",
    "    ['chr1', '65565', '65567', 'OR4F5-202', '0', '+', 'start_codon'],\n",
    "    ['chr1', '69037', '71585', 'OR4F5-202', '0', '+', 'exon'],\n",
    "    ['chr1', '70006', '70008', 'OR4F5-202', '0', '+', 'stop_codon'],\n",
    "    ['chr1', '65419', '65433', 'OR4F5-202', '0', '+', 'UTR'],\n",
    "    ['chr1', '65520', '65564', 'OR4F5-202', '0', '+', 'UTR'],\n",
    "    ['chr1', '70006', '71585', 'OR4F5-202', '0', '+', 'UTR']\n",
    "]\n",
    "getPosBed_Plus(arrTestP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#part1.2: get position bed from transcript bed -\n",
    "# v0.2 add stop codon\n",
    "def getPosBed_Minus(transBed):\n",
    "    arrTest=transBed\n",
    "    \n",
    "    ## init paras\n",
    "    arrExon=[]\n",
    "    arrUTR=[]\n",
    "    start_codon=[] #one\n",
    "    stop_codon=[] #one\n",
    "    #\n",
    "    for tmp in arrTest:\n",
    "        #arr3=arrTest[i];\n",
    "        if tmp[-1]=='exon':\n",
    "            arrExon.append(tmp)\n",
    "        elif tmp[-1]=='UTR':\n",
    "            arrUTR.append(tmp)\n",
    "        elif tmp[-1]=='start_codon':\n",
    "            start_codon=tmp;\n",
    "        elif tmp[-1]=='stop_codon':\n",
    "            stop_codon=tmp;\n",
    "        #print(arrTest[i])\n",
    "    #print(len(arrExon), len(arrUTR))\n",
    "\n",
    "    \n",
    "    ## get 3UTR, maybe\n",
    "    UTR3=[]\n",
    "    UTR5=[]\n",
    "    if len(arrUTR)>0:\n",
    "        for tmp in arrUTR:\n",
    "            if len(start_codon)>0 and tmp[1]>start_codon[2]:\n",
    "                tmp[-1]=\"UTR5\"\n",
    "                UTR5.append(tmp)\n",
    "            elif len(stop_codon)>0 and tmp[2]<=stop_codon[2]:\n",
    "                tmp[-1]=\"UTR3\"\n",
    "                UTR3.append(tmp)\n",
    "    #print(len(UTR5), len(UTR3))\n",
    "    \n",
    "    \n",
    "    ## get e3UTR, maybe\n",
    "    eUTR3=[]\n",
    "    if len(UTR3)>0:\n",
    "        eUTR3=UTR3[-1].copy()\n",
    "        #\n",
    "        eUTR3[2]= str( int(eUTR3[1])-1 )\n",
    "        eUTR3[1]= str( int(eUTR3[1])-1000 )\n",
    "        eUTR3[6]='extended3UTR'\n",
    "        #print(UTR3[-1])\n",
    "        #print(eUTR3)\n",
    "        \n",
    "    ##get arrIntron，must\n",
    "    arrIntron=[]\n",
    "    #print(arrTest[0], '\\n')\n",
    "    intron=[]\n",
    "    if len(arrExon)>1:\n",
    "        for i in range(len(arrExon)):\n",
    "            tmp=arrExon[i]\n",
    "            if len(intron)==0: #first exon\n",
    "                intron=tmp.copy()\n",
    "                intron[2]= str( int(intron[1])-1 )\n",
    "                continue;\n",
    "            else:\n",
    "                #第二个时\n",
    "                intron[1]=str( int(tmp[2])+1 )\n",
    "                intron[-1]=\"intron\"\n",
    "                arrIntron.append(intron);\n",
    "                #\n",
    "                intron=tmp.copy();\n",
    "                intron[2]= str( int(intron[1])-1 )\n",
    "    #\n",
    "    #print('arrExon=',arrExon)\n",
    "    #print('arrIntron=',arrIntron)\n",
    "\n",
    "    ##get TSS，must\n",
    "    TSS=arrTest[0].copy()\n",
    "    pos=int(arrTest[0][2])\n",
    "    TSS[2]=str( pos+10 )\n",
    "    TSS[1]=str( pos-10 )\n",
    "    TSS[-1]=\"TSS\"\n",
    "    #print(arrTest[0])\n",
    "    #TSS\n",
    "    \n",
    "    ##get Promoter，must\n",
    "    Promoter=arrTest[0].copy()\n",
    "    Promoter[1]=str( pos+11 )\n",
    "    Promoter[2]=str( pos+250 )\n",
    "    Promoter[-1]='Promoter'\n",
    "    #print(arrTest[0])\n",
    "    #Promoter\n",
    "    \n",
    "    #get PA must\n",
    "    PA=arrTest[0].copy()\n",
    "    pos2=int(arrTest[0][1])\n",
    "    PA[1]=str( pos2-10 )\n",
    "    PA[2]=str( pos2+10 )\n",
    "    PA[-1]='PA'\n",
    "    \n",
    "    # 合并\n",
    "    saveSet=[]\n",
    "    saveSet.extend(UTR3) #many\n",
    "    saveSet.append(eUTR3) #1\n",
    "\n",
    "    saveSet.extend(arrExon) #many\n",
    "    saveSet.extend(arrIntron) #many\n",
    "    saveSet.extend(UTR5) #many\n",
    "\n",
    "    saveSet.append(TSS) #1\n",
    "    saveSet.append(Promoter) #1\n",
    "    saveSet.append(arrTest[0]) #1\n",
    "    \n",
    "    saveSet.append(stop_codon)\n",
    "    saveSet.append(start_codon)\n",
    "    \n",
    "    saveSet.append(PA)\n",
    "\n",
    "    return saveSet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['chr1', '879583', '880076', 'NOC2L-201', '0', '-', 'UTR3'],\n",
       " ['chr1', '878583', '879582', 'NOC2L-201', '0', '-', 'extended3UTR'],\n",
       " ['chr1', '894595', '894636', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '894309', '894461', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '892479', '892653', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '892274', '892405', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '891475', '891595', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '891303', '891393', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '889384', '889462', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '889162', '889272', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '888555', '888668', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '887792', '887980', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '887380', '887519', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '886507', '886618', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '883870', '883983', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '883511', '883612', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '881782', '881925', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '881553', '881666', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '880898', '881033', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '880437', '880526', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '879583', '880180', 'NOC2L-201', '0', '-', 'exon'],\n",
       " ['chr1', '894462', '894594', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '892654', '894308', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '892406', '892478', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '891596', '892273', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '891394', '891474', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '889463', '891302', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '889273', '889383', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '888669', '889161', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '887981', '888554', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '887520', '887791', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '886619', '887379', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '883984', '886506', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '883613', '883869', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '881926', '883510', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '881667', '881781', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '881034', '881552', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '880527', '880897', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '880181', '880436', 'NOC2L-201', '0', '-', 'intron'],\n",
       " ['chr1', '894621', '894636', 'NOC2L-201', '0', '-', 'UTR5'],\n",
       " ['chr1', '894626', '894646', 'NOC2L-201', '0', '-', 'TSS'],\n",
       " ['chr1', '894647', '894886', 'NOC2L-201', '0', '-', 'Promoter'],\n",
       " ['chr1', '879583', '894636', 'NOC2L-201', '0', '-', 'transcript'],\n",
       " ['chr1', '880074', '880076', 'NOC2L-201', '0', '-', 'stop_codon'],\n",
       " ['chr1', '894618', '894620', 'NOC2L-201', '0', '-', 'start_codon'],\n",
       " ['chr1', '879573', '879593', 'NOC2L-201', '0', '-', 'PA']]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#test\n",
    "arrTestM=[\n",
    "    ['chr1', '879583', '894636', 'NOC2L-201', '0', '-', 'transcript'],\n",
    "    ['chr1', '894595', '894636', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '894618', '894620', 'NOC2L-201', '0', '-', 'start_codon'],\n",
    "    ['chr1', '894309', '894461', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '892479', '892653', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '892274', '892405', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '891475', '891595', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '891303', '891393', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '889384', '889462', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '889162', '889272', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '888555', '888668', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '887792', '887980', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '887380', '887519', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '886507', '886618', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '883870', '883983', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '883511', '883612', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '881782', '881925', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '881553', '881666', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '880898', '881033', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '880437', '880526', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '879583', '880180', 'NOC2L-201', '0', '-', 'exon'],\n",
    "    ['chr1', '880074', '880076', 'NOC2L-201', '0', '-', 'stop_codon'],\n",
    "    ['chr1', '894621', '894636', 'NOC2L-201', '0', '-', 'UTR'],\n",
    "    ['chr1', '879583', '880076', 'NOC2L-201', '0', '-', 'UTR']\n",
    "]\n",
    "getPosBed_Minus(arrTestM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10000 0.37 second elapsed.\n",
      "20000 0.66 second elapsed.\n",
      "30000 0.95 second elapsed.\n",
      "40000 1.24 second elapsed.\n",
      "50000 1.53 second elapsed.\n",
      "60000 1.81 second elapsed.\n",
      "70000 2.1 second elapsed.\n",
      "80000 2.39 second elapsed.\n",
      "90000 2.67 second elapsed.\n",
      "100000 2.96 second elapsed.\n",
      "110000 3.25 second elapsed.\n",
      "120000 3.54 second elapsed.\n",
      "130000 3.83 second elapsed.\n",
      "140000 4.11 second elapsed.\n",
      "150000 4.41 second elapsed.\n",
      "160000 4.69 second elapsed.\n",
      "170000 4.99 second elapsed.\n",
      "180000 5.27 second elapsed.\n",
      "190000 5.56 second elapsed.\n",
      "200000 5.85 second elapsed.\n",
      "210000 6.14 second elapsed.\n",
      "220000 6.43 second elapsed.\n",
      "230000 6.72 second elapsed.\n",
      "240000 7.0 second elapsed.\n",
      "250000 7.29 second elapsed.\n",
      "260000 7.58 second elapsed.\n",
      "270000 7.88 second elapsed.\n",
      "280000 8.17 second elapsed.\n",
      "290000 8.46 second elapsed.\n",
      "300000 8.75 second elapsed.\n",
      "310000 9.05 second elapsed.\n",
      "320000 9.51 second elapsed.\n",
      "330000 9.79 second elapsed.\n",
      "340000 10.08 second elapsed.\n",
      "350000 10.37 second elapsed.\n",
      "360000 10.66 second elapsed.\n",
      "370000 10.94 second elapsed.\n",
      "380000 11.23 second elapsed.\n",
      "390000 11.63 second elapsed.\n",
      "400000 11.91 second elapsed.\n",
      "410000 12.2 second elapsed.\n",
      "420000 12.49 second elapsed.\n",
      "430000 12.77 second elapsed.\n",
      "440000 13.06 second elapsed.\n",
      "450000 13.34 second elapsed.\n",
      "460000 13.63 second elapsed.\n",
      "470000 13.92 second elapsed.\n",
      "480000 14.22 second elapsed.\n",
      "490000 14.5 second elapsed.\n",
      "500000 14.79 second elapsed.\n",
      "510000 15.07 second elapsed.\n",
      "520000 15.36 second elapsed.\n",
      "530000 15.64 second elapsed.\n",
      "540000 15.92 second elapsed.\n",
      "550000 16.21 second elapsed.\n",
      "560000 16.49 second elapsed.\n",
      "570000 16.78 second elapsed.\n",
      "580000 17.06 second elapsed.\n",
      "590000 17.35 second elapsed.\n",
      "600000 17.64 second elapsed.\n",
      "610000 17.93 second elapsed.\n",
      "620000 18.21 second elapsed.\n",
      "630000 18.5 second elapsed.\n",
      "640000 18.78 second elapsed.\n",
      "650000 19.08 second elapsed.\n",
      "660000 19.39 second elapsed.\n",
      "670000 19.68 second elapsed.\n",
      "680000 19.96 second elapsed.\n",
      "690000 20.24 second elapsed.\n",
      "700000 20.53 second elapsed.\n",
      "710000 20.81 second elapsed.\n",
      "720000 21.1 second elapsed.\n",
      "730000 21.39 second elapsed.\n",
      "740000 21.69 second elapsed.\n",
      "750000 21.98 second elapsed.\n",
      "760000 22.26 second elapsed.\n",
      "770000 22.55 second elapsed.\n",
      "780000 22.84 second elapsed.\n",
      "790000 23.13 second elapsed.\n",
      "800000 23.42 second elapsed.\n",
      "810000 23.71 second elapsed.\n",
      "820000 24.0 second elapsed.\n",
      "830000 24.28 second elapsed.\n",
      "840000 24.57 second elapsed.\n",
      "850000 24.88 second elapsed.\n",
      "860000 25.18 second elapsed.\n",
      "870000 25.46 second elapsed.\n",
      "880000 25.75 second elapsed.\n",
      "890000 26.03 second elapsed.\n",
      "900000 26.33 second elapsed.\n",
      "910000 26.62 second elapsed.\n",
      "920000 26.92 second elapsed.\n",
      "930000 27.21 second elapsed.\n",
      "940000 27.5 second elapsed.\n",
      "950000 27.79 second elapsed.\n",
      "960000 28.08 second elapsed.\n",
      "970000 28.38 second elapsed.\n",
      "980000 28.67 second elapsed.\n",
      "990000 28.96 second elapsed.\n",
      "1000000 29.25 second elapsed.\n",
      "1010000 29.57 second elapsed.\n",
      "1020000 29.87 second elapsed.\n",
      "1030000 30.16 second elapsed.\n",
      "1040000 30.45 second elapsed.\n",
      "1050000 30.75 second elapsed.\n",
      "1060000 31.04 second elapsed.\n",
      "1070000 31.33 second elapsed.\n",
      "1080000 31.62 second elapsed.\n",
      "1090000 31.92 second elapsed.\n",
      "1100000 32.21 second elapsed.\n",
      "1110000 32.5 second elapsed.\n",
      "1120000 32.79 second elapsed.\n",
      "1130000 33.09 second elapsed.\n",
      "1140000 33.38 second elapsed.\n",
      "1150000 33.67 second elapsed.\n",
      "1160000 33.95 second elapsed.\n",
      "1170000 34.24 second elapsed.\n",
      "1180000 34.53 second elapsed.\n",
      "1190000 34.82 second elapsed.\n",
      "1200000 35.1 second elapsed.\n",
      "1210000 35.4 second elapsed.\n",
      "1220000 35.69 second elapsed.\n",
      "1230000 35.98 second elapsed.\n",
      "1240000 36.27 second elapsed.\n",
      "1250000 36.56 second elapsed.\n",
      "1260000 36.85 second elapsed.\n",
      "1270000 37.13 second elapsed.\n",
      "1280000 37.42 second elapsed.\n",
      "1290000 37.72 second elapsed.\n",
      "1300000 38.01 second elapsed.\n",
      "1310000 38.3 second elapsed.\n",
      "1320000 38.58 second elapsed.\n",
      "1330000 38.86 second elapsed.\n",
      "1340000 39.15 second elapsed.\n",
      "1350000 39.43 second elapsed.\n",
      "1360000 39.83 second elapsed.\n",
      "1370000 40.12 second elapsed.\n",
      "1380000 40.41 second elapsed.\n",
      "1390000 40.69 second elapsed.\n",
      "1400000 40.97 second elapsed.\n",
      "1410000 41.26 second elapsed.\n",
      "1420000 41.55 second elapsed.\n",
      "1430000 41.84 second elapsed.\n",
      "1440000 42.13 second elapsed.\n",
      "1450000 42.42 second elapsed.\n",
      "1460000 42.71 second elapsed.\n",
      "1470000 43.0 second elapsed.\n",
      "1480000 43.29 second elapsed.\n",
      "1490000 43.58 second elapsed.\n",
      "1500000 43.87 second elapsed.\n",
      "1510000 44.15 second elapsed.\n",
      "1520000 44.45 second elapsed.\n",
      "1530000 44.74 second elapsed.\n",
      "1540000 45.04 second elapsed.\n",
      "1550000 45.32 second elapsed.\n",
      "1560000 45.61 second elapsed.\n",
      "1570000 45.9 second elapsed.\n",
      "1580000 46.19 second elapsed.\n",
      "1590000 46.48 second elapsed.\n",
      "1600000 46.78 second elapsed.\n",
      "1610000 47.06 second elapsed.\n",
      "1620000 47.36 second elapsed.\n",
      "1630000 47.65 second elapsed.\n",
      "1640000 47.94 second elapsed.\n",
      "1650000 48.23 second elapsed.\n",
      "1660000 48.52 second elapsed.\n",
      "1670000 48.83 second elapsed.\n",
      "1680000 49.12 second elapsed.\n",
      "1690000 49.42 second elapsed.\n",
      "1700000 49.89 second elapsed.\n",
      "1710000 50.18 second elapsed.\n",
      "1720000 50.48 second elapsed.\n",
      "1730000 50.76 second elapsed.\n",
      "1740000 51.06 second elapsed.\n",
      "1750000 51.36 second elapsed.\n",
      "1760000 51.65 second elapsed.\n",
      "1770000 51.94 second elapsed.\n",
      "1780000 52.24 second elapsed.\n",
      "1790000 52.53 second elapsed.\n",
      "1800000 52.82 second elapsed.\n",
      "1810000 53.16 second elapsed.\n",
      "1820000 53.45 second elapsed.\n",
      "1830000 53.75 second elapsed.\n",
      "1840000 54.04 second elapsed.\n",
      "1850000 54.33 second elapsed.\n",
      "1860000 54.62 second elapsed.\n",
      "1870000 54.9 second elapsed.\n",
      "1880000 55.2 second elapsed.\n",
      "1890000 55.49 second elapsed.\n",
      "1900000 55.78 second elapsed.\n",
      "1910000 56.08 second elapsed.\n",
      "1920000 56.37 second elapsed.\n",
      "1930000 56.66 second elapsed.\n",
      "1940000 56.96 second elapsed.\n",
      "1950000 57.26 second elapsed.\n",
      "1960000 57.57 second elapsed.\n",
      "==main end== i= 1967895 ; totalLine= 2782645\n"
     ]
    }
   ],
   "source": [
    "#genomic location based on GTF\n",
    "import re, time\n",
    "\n",
    "gtf='/home/wangjl/data/ref/hg19/gencode.v30lift37.annotation.gtf'\n",
    "fr=open(gtf,'r',encoding=\"utf-8\")\n",
    "\n",
    "start_time=time.time()\n",
    "\n",
    "arrFile={}\n",
    "dirPath_beds=\"/home/wangjl/data/apa/191111Figure/f2/positionPY/\"\n",
    "arrFile['UTR3']=open(dirPath_beds+\"UTR3.bed\", 'w')\n",
    "arrFile['extended3UTR']=open(dirPath_beds+\"extended3UTR.bed\", 'w')\n",
    "arrFile['exon']=open(dirPath_beds+\"exon.bed\", 'w')\n",
    "arrFile['intron']=open(dirPath_beds+\"intron.bed\", 'w')\n",
    "arrFile['UTR5']=open(dirPath_beds+\"UTR5.bed\", 'w')\n",
    "arrFile['TSS']=open(dirPath_beds+\"TSS.bed\", 'w')\n",
    "arrFile['Promoter']=open(dirPath_beds+\"Promoter.bed\", 'w')\n",
    "arrFile['transcript']=open(dirPath_beds+\"transcript.bed\", 'w')\n",
    "#\n",
    "arrFile['start_codon']=open(dirPath_beds+\"start_codon.bed\", 'w')\n",
    "arrFile['stop_codon']=open(dirPath_beds+\"stop_codon.bed\", 'w')\n",
    "arrFile['PA']=open(dirPath_beds+\"PA.bed\", 'w')\n",
    "\n",
    "#\n",
    "i=0\n",
    "db=[]\n",
    "totalLine=0\n",
    "while True:\n",
    "    totalLine+=1\n",
    "    line=fr.readline()\n",
    "    if not line: #last line\n",
    "        break\n",
    "    \n",
    "    line=line.strip()\n",
    "    \n",
    "    # filter: not in use now\n",
    "    if not re.search('protein_coding', line):\n",
    "        #continue;\n",
    "        pass\n",
    "    \n",
    "    # filter: discard comment lines #\n",
    "    if line.startswith(\"#\"):\n",
    "        continue;\n",
    "\n",
    "    #\n",
    "    arr=re.split('\\t', line) ##['chr1', 'HAVANA', 'transcript', '65419', '71585', '.', '+', '.', ...]    \n",
    "    # filter out cds\n",
    "    if arr[2]=='CDS' or arr[2]=='gene':\n",
    "        continue;\n",
    "    #\n",
    "        \n",
    "    #for debug\n",
    "    i+=1;\n",
    "    if i>3000:\n",
    "        #break;\n",
    "        pass\n",
    "    \n",
    "    if i%10000==0:\n",
    "        print(i, round(time.time()-start_time,2), 'second elapsed.')\n",
    "    \n",
    "    ##############\n",
    "    # get a transcript\n",
    "    arr2=re.split('; ', arr[8])\n",
    "    #print( arr2 )\n",
    "    #['gene_id \"ENSG00000186092.6_4\"', 'transcript_id \"ENST00000641515.2_2\"', \n",
    "    #'gene_type \"protein_coding\"', 'gene_name \"OR4F5\"', 'transcript_type \"protein_coding\"', \n",
    "    #'transcript_name \"OR4F5-202\"', 'level 2', 'protein_id \"ENSP00000493376.2\"', 'tag \"RNA_Seq_supported_partial\"', \n",
    "    #'tag \"basic\"', 'havana_gene \"OTTHUMG00000001094.4_4\"', 'havana_transcript \"OTTHUMT00000003223.4_2\"', 'remap_num_mappings 1', \n",
    "    #'remap_status \"full_contig\"', 'remap_target_status \"new\";']\n",
    "\n",
    "    dictG={}\n",
    "    for item in arr2:\n",
    "        (k,v)=re.split(' ',item)\n",
    "        dictG[k]=v.strip('\"')\n",
    "    #rint('dictG=',dictG)\n",
    "    if 'transcript_name' in dictG:\n",
    "        transcript_name=dictG['transcript_name']\n",
    "    else:\n",
    "        transcript_name=\"\"\n",
    "        #continue;\n",
    "    #\n",
    "    arrN=[arr[0], arr[3], arr[4], transcript_name, \"0\", arr[6], arr[2] ]\n",
    "        \n",
    "    #save to file, according to type(start, end, 3UTR, e3UTR, exon, intron, 5UTR, TSS, promoter) and strand.\n",
    "    if arr[2]=='transcript':\n",
    "        if len(db)>0:\n",
    "            #print('i=', i, '; len(db)=', len(db))            \n",
    "            #for j in db:\n",
    "            #    print(j) # ['chr1', '65419', '71585', 'OR4F5-202', '0', '+']\n",
    "            #    pass\n",
    "            \n",
    "            #get strand\n",
    "            if db[0][5]=='-':\n",
    "                saveSet=getPosBed_Minus(db)\n",
    "            else:\n",
    "                saveSet=getPosBed_Plus(db)\n",
    "            #save to file\n",
    "            #print('; len(saveSet)=',len(saveSet))\n",
    "            for j in saveSet:\n",
    "                if len(j)>0:\n",
    "                    arrFile[j[-1]].write( '\\t'.join(j)+'\\n' ) # ['chr1', '65419', '71585', 'OR4F5-202', '0', '+']\n",
    "                #pass\n",
    "            \n",
    "        \n",
    "        # clear and next transcript\n",
    "        db=[]\n",
    "        db.append(arrN)\n",
    "        #print('=*'*20)\n",
    "    else:\n",
    "        db.append(arrN)\n",
    "    \n",
    "    #print(i, arr[0:8], len(arr))\n",
    "    \n",
    "fr.close()\n",
    "\n",
    "for f in arrFile:\n",
    "    arrFile[f].close()\n",
    "print('==main end== i=',i, '; totalLine=',totalLine) #57s 2782645 lines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# part 2 read bed files to memory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# position (as para name)= chr:strand { 12: gene1, 15: gene2} \n",
    "\n",
    "import re,time\n",
    "def readBed(bedName, verbose=False):\n",
    "    print(\"File name: \",bedName)\n",
    "    start_time=time.time()\n",
    "    fr=open(\"/home/wangjl/data/apa/191111Figure/f2/positionPY/\"+bedName+\".bed\", 'r')\n",
    "\n",
    "    i=0\n",
    "    mem={}\n",
    "    for lineR in fr.readlines():\n",
    "        i+=1\n",
    "        if i>1000:\n",
    "            #break;\n",
    "            pass\n",
    "        if i%1e5==0:\n",
    "            print('\\t==> ',i, round(time.time()-start_time,2), 'second elapsed.')\n",
    "        \n",
    "        line=lineR.strip()\n",
    "        arr=re.split('\\t', line)\n",
    "        #print(i, arr)\n",
    "\n",
    "        chrStrand=arr[0]+\":\"+arr[5]\n",
    "        if chrStrand not in mem:\n",
    "            mem[chrStrand]={}\n",
    "        for j in range( int(arr[1]), int(arr[2])+1):\n",
    "            mem[chrStrand][j]=arr[3]\n",
    "\n",
    "    fr.close()\n",
    "    print('\\t==end==, i=', i, bedName)\n",
    "    if verbose:\n",
    "        for chrStrand in mem:\n",
    "            print(chrStrand, len( mem[chrStrand]) )\n",
    "    return mem;\n",
    "\n",
    "#test\n",
    "#extended3UTR=readBed(\"extended3UTR\", True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# part 3New(use less memory) annotate location and transcript name for each given point"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==end==,i= 742288\n"
     ]
    }
   ],
   "source": [
    "# add 2 cols\n",
    "posFile=\"/data/jinwf/wangjl/apa/190705PAS/bed/pasPostions_siteAndIncludes_PY.bed\"\n",
    "outputFile=\"/data/jinwf/wangjl/apa/190705PAS/bed/pasPostions_Location_transcriptName_PY.bed\"\n",
    "\n",
    "def extend2Column():\n",
    "    fr=open(posFile,'r',encoding=\"utf-8\")\n",
    "    fw=open(outputFile, 'w')\n",
    "\n",
    "    i=0\n",
    "    for lineR in fr.readlines():\n",
    "        i+=1\n",
    "        line=lineR.strip()\n",
    "        arr=re.split('\\t',line)\n",
    "        # if only 6 cols, then add 2 last cols\n",
    "        if len(arr)<=6:\n",
    "            arr.append('intergenic') #region\n",
    "            arr.append('.') #gene names\n",
    "        fw.write('\\t'.join(arr)+\"\\n\")    \n",
    "    fr.close();\n",
    "    fw.close();\n",
    "    print('==end==,i=',i);\n",
    "\n",
    "extend2Column();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# update position genomic region\n",
    "def updateRegion(location, outputFile):\n",
    "    # read the DB;\n",
    "    posDB=readBed(location)\n",
    "    \n",
    "    #middle file;\n",
    "    tmpFile=\"tmp_123456789.txt\"\n",
    "\n",
    "    fr=open(outputFile, 'r')\n",
    "    fw=open(tmpFile, 'w')\n",
    "    i=0;\n",
    "    j=0;\n",
    "    for lineR in fr.readlines():\n",
    "        i+=1;\n",
    "        if i>100:\n",
    "            #break;\n",
    "            pass;\n",
    "        line=lineR.strip()\n",
    "        arr=re.split('\\t',line)\n",
    "\n",
    "        # jump if have annotated, so first annotate prioritized regions.\n",
    "        if arr[6]!='intergenic':\n",
    "            fw.write('\\t'.join(arr)+\"\\n\");\n",
    "            continue;\n",
    "\n",
    "        # add annotate info, refer to PAS DB\n",
    "        chrStrand=arr[0]+':'+arr[5]\n",
    "        pos=int(arr[1])\n",
    "        if arr[5]==\"-\":\n",
    "            pos+=1\n",
    "        if chrStrand in posDB:\n",
    "            if pos in posDB[chrStrand]:\n",
    "                transcriptName=posDB[chrStrand][pos]\n",
    "                # if find, over, and write down\n",
    "                arr[6]=location;\n",
    "                arr[7]=transcriptName\n",
    "                j+=1;\n",
    "        fw.write('\\t'.join(arr)+\"\\n\");\n",
    "    \n",
    "    fr.close();\n",
    "    fw.close();\n",
    "    print(location, 'change=',j, ', total=',i, round(time.time()- start_time, 2),'seconds elapsed.')\n",
    "\n",
    "    #change file name\n",
    "    os.rename(tmpFile, outputFile);\n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==end==,i= 742288\n",
      "==================== PA\n",
      "File name:  PA\n",
      "\t==>  100000 0.79 second elapsed.\n",
      "\t==>  200000 1.31 second elapsed.\n",
      "\t==end==, i= 210424 PA\n",
      "PA change= 27726 , total= 742288 4.82 seconds elapsed.\n",
      "-------------------- 3.59 s\n",
      "==================== UTR3\n",
      "File name:  UTR3\n",
      "\t==>  100000 9.48 second elapsed.\n",
      "\t==end==, i= 154332 UTR3\n",
      "UTR3 change= 73214 , total= 742288 20.43 seconds elapsed.\n",
      "-------------------- 16.25 s\n",
      "==================== extended3UTR\n",
      "File name:  extended3UTR\n",
      "\t==end==, i= 78919 extended3UTR\n",
      "extended3UTR change= 13083 , total= 742288 34.43 seconds elapsed.\n",
      "-------------------- 14.03 s\n",
      "==================== exon\n",
      "File name:  exon\n",
      "\t==>  100000 4.23 second elapsed.\n",
      "\t==>  200000 8.46 second elapsed.\n",
      "\t==>  300000 12.02 second elapsed.\n",
      "\t==>  400000 18.41 second elapsed.\n",
      "\t==>  500000 22.35 second elapsed.\n",
      "\t==>  600000 26.34 second elapsed.\n",
      "\t==>  700000 30.46 second elapsed.\n",
      "\t==>  800000 34.06 second elapsed.\n",
      "\t==>  900000 37.83 second elapsed.\n",
      "\t==>  1000000 41.53 second elapsed.\n",
      "\t==>  1100000 46.15 second elapsed.\n",
      "\t==>  1200000 50.99 second elapsed.\n",
      "\t==end==, i= 1283655 exon\n",
      "exon change= 56091 , total= 742288 93.37 seconds elapsed.\n",
      "-------------------- 61.48 s\n",
      "==================== intron\n",
      "File name:  intron\n",
      "\t==>  100000 75.94 second elapsed.\n",
      "\t==>  200000 214.58 second elapsed.\n",
      "\t==>  300000 372.66 second elapsed.\n",
      "\t==>  400000 535.84 second elapsed.\n",
      "\t==>  500000 687.03 second elapsed.\n",
      "\t==>  600000 785.48 second elapsed.\n",
      "\t==>  700000 892.93 second elapsed.\n",
      "\t==>  800000 984.04 second elapsed.\n",
      "\t==>  900000 1052.98 second elapsed.\n",
      "\t==>  1000000 1123.14 second elapsed.\n",
      "\t==end==, i= 1073231 intron\n",
      "intron change= 261757 , total= 742288 1298.49 seconds elapsed.\n",
      "-------------------- 1301.14 s\n",
      "==================== UTR5\n",
      "File name:  UTR5\n",
      "\t==>  100000 2.75 second elapsed.\n",
      "\t==end==, i= 151867 UTR5\n",
      "UTR5 change= 0 , total= 742288 1403.18 seconds elapsed.\n",
      "-------------------- 5.49 s\n",
      "==================== TSS\n",
      "File name:  TSS\n",
      "\t==>  100000 0.64 second elapsed.\n",
      "\t==>  200000 1.13 second elapsed.\n",
      "\t==end==, i= 210424 TSS\n",
      "TSS change= 49 , total= 742288 1406.13 seconds elapsed.\n",
      "-------------------- 2.82 s\n",
      "==================== Promoter\n",
      "File name:  Promoter\n",
      "\t==>  100000 4.04 second elapsed.\n",
      "\t==>  200000 7.8 second elapsed.\n",
      "\t==end==, i= 210424 Promoter\n",
      "Promoter change= 1590 , total= 742288 1416.04 seconds elapsed.\n",
      "-------------------- 10.38 s\n",
      "==end==\n"
     ]
    }
   ],
   "source": [
    "# v0.2\n",
    "#task1: OK;\n",
    "# depend on readBed() before;\n",
    "\n",
    "import re, time, os\n",
    "# location, transcript name;\n",
    "\n",
    "\n",
    "def main():\n",
    "    #1. add 2 cols\n",
    "    extend2Column();\n",
    "    \n",
    "    #2. add annotation\n",
    "    locatonArr=['PA', 'UTR3','extended3UTR', 'exon','intron','UTR5','TSS','Promoter']\n",
    "    for location in locatonArr:\n",
    "        t1=time.time();\n",
    "        print('='*20, location)\n",
    "        updateRegion(location, outputFile)\n",
    "        print('-'*20, round(time.time()-t1,2),'s' )\n",
    "#\n",
    "\n",
    "start_time=time.time();\n",
    "main();\n",
    "print('==end==');"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## for DB3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==end==,i= 311595\n"
     ]
    }
   ],
   "source": [
    "# for DB3\n",
    "\n",
    "# add cols\n",
    "posFile=   \"/home/wangjl/data/apa/190610APA/02_validate/human.PAS.txt\"\n",
    "outputFile=\"/data/jinwf/wangjl/apa/190705PAS/bed/PolyA_DB3_location_transcriptName.bed\"\n",
    "#chr1    564599  564599  chr1:564599:+   0       +       exon    MTND1P23-201\n",
    "\n",
    "\n",
    "def extend2Column2(posFile2, outputFile2):\n",
    "    fr=open(posFile2,'r',encoding=\"utf-8\")\n",
    "    fw=open(outputFile2, 'w')\n",
    "\n",
    "    i=0\n",
    "    for lineR in fr.readlines():\n",
    "        i+=1\n",
    "        if i==1:\n",
    "            continue;\n",
    "        if i>10:\n",
    "            #break;\n",
    "            pass;\n",
    "    \n",
    "        line=lineR.strip()\n",
    "        arr=re.split('\\t',line)\n",
    "        \n",
    "        arrN=[arr[1], arr[2],arr[2],arr[0], \"0\", arr[3]]\n",
    "        ##chr10\t157706\t157706\tchr10:157706:-\t5\t-\n",
    "        \n",
    "        # add 2 cols if only 6 cols.\n",
    "        if len(arrN)<=6:\n",
    "            arrN.append('intergenic') #regions\n",
    "            arrN.append('.') #gene symbols\n",
    "        fw.write('\\t'.join(arrN)+\"\\n\")    \n",
    "    fr.close();\n",
    "    fw.close();\n",
    "    print('==end==,i=',i);\n",
    "\n",
    "extend2Column2(posFile, outputFile);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==end==,i= 311595\n",
      "==================== PA\n",
      "File name:  PA\n",
      "\t==>  100000 0.52 second elapsed.\n",
      "\t==>  200000 1.03 second elapsed.\n",
      "\t==end==, i= 210424 PA\n",
      "PA change= 36015 , total= 311594 2.74 seconds elapsed.\n",
      "-------------------- 1.91 s\n",
      "==================== UTR3\n",
      "File name:  UTR3\n",
      "\t==>  100000 8.34 second elapsed.\n",
      "\t==end==, i= 154332 UTR3\n",
      "UTR3 change= 74951 , total= 311594 15.85 seconds elapsed.\n",
      "-------------------- 13.74 s\n",
      "==================== extended3UTR\n",
      "File name:  extended3UTR\n",
      "\t==end==, i= 78919 extended3UTR\n",
      "extended3UTR change= 14991 , total= 311594 28.43 seconds elapsed.\n",
      "-------------------- 12.49 s\n",
      "==================== exon\n",
      "File name:  exon\n",
      "\t==>  100000 4.38 second elapsed.\n",
      "\t==>  200000 8.68 second elapsed.\n",
      "\t==>  300000 12.25 second elapsed.\n",
      "\t==>  400000 16.6 second elapsed.\n",
      "\t==>  500000 20.56 second elapsed.\n",
      "\t==>  600000 24.55 second elapsed.\n",
      "\t==>  700000 28.24 second elapsed.\n",
      "\t==>  800000 31.9 second elapsed.\n",
      "\t==>  900000 35.69 second elapsed.\n",
      "\t==>  1000000 39.4 second elapsed.\n",
      "\t==>  1100000 43.09 second elapsed.\n",
      "\t==>  1200000 46.86 second elapsed.\n",
      "\t==end==, i= 1283655 exon\n",
      "exon change= 20256 , total= 311594 79.93 seconds elapsed.\n",
      "-------------------- 53.0 s\n",
      "==================== intron\n",
      "File name:  intron\n",
      "\t==>  100000 74.37 second elapsed.\n",
      "\t==>  200000 181.05 second elapsed.\n",
      "\t==>  300000 334.04 second elapsed.\n",
      "\t==>  400000 458.39 second elapsed.\n",
      "\t==>  500000 595.09 second elapsed.\n",
      "\t==>  600000 695.47 second elapsed.\n",
      "\t==>  700000 797.07 second elapsed.\n",
      "\t==>  800000 881.45 second elapsed.\n",
      "\t==>  900000 941.34 second elapsed.\n",
      "\t==>  1000000 1012.18 second elapsed.\n",
      "\t==end==, i= 1073231 intron\n",
      "intron change= 79931 , total= 311594 1168.11 seconds elapsed.\n",
      "-------------------- 1167.25 s\n",
      "==================== UTR5\n",
      "File name:  UTR5\n",
      "\t==>  100000 2.02 second elapsed.\n",
      "\t==end==, i= 151867 UTR5\n",
      "UTR5 change= 0 , total= 311594 1252.89 seconds elapsed.\n",
      "-------------------- 3.87 s\n",
      "==================== TSS\n",
      "File name:  TSS\n",
      "\t==>  100000 0.65 second elapsed.\n",
      "\t==>  200000 1.13 second elapsed.\n",
      "\t==end==, i= 210424 TSS\n",
      "TSS change= 26 , total= 311594 1254.96 seconds elapsed.\n",
      "-------------------- 1.91 s\n",
      "==================== Promoter\n",
      "File name:  Promoter\n",
      "\t==>  100000 3.95 second elapsed.\n",
      "\t==>  200000 7.56 second elapsed.\n",
      "\t==end==, i= 210424 Promoter\n",
      "Promoter change= 606 , total= 311594 1263.7 seconds elapsed.\n",
      "-------------------- 9.39 s\n",
      "==end==\n"
     ]
    }
   ],
   "source": [
    "#######################\n",
    "# ===> annotate site for polyA_DB3\n",
    "#######################\n",
    "def main2():\n",
    "    #1. add cols\n",
    "    extend2Column2(posFile, outputFile);\n",
    "    \n",
    "    #2. add annotation\n",
    "    locatonArr=['PA', 'UTR3','extended3UTR', 'exon','intron','UTR5','TSS','Promoter']\n",
    "    for location in locatonArr:\n",
    "        t1=time.time();\n",
    "        print('='*20, location)\n",
    "        updateRegion(location, outputFile)\n",
    "        print('-'*20, round(time.time()-t1,2),'s' )\n",
    "#\n",
    "\n",
    "start_time=time.time();\n",
    "main2();\n",
    "print('==end==');"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
